File: java/104maxdepthofbinarytree/104.二叉树的最大深度.java
/*
 * @lc app=leetcode.cn id=104 lang=java
 *
 * [104] 二叉树的最大深度
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int maxDepth(TreeNode root) {
        return getDepth(root, 0);
    }
    
    public int getDepth(TreeNode node, int depth) {
        if (node == null) {
            return depth;
        }
        depth++;
        int leftDepth = getDepth(node.left, depth);
        int rightDepth = getDepth(node.right, depth);
        if (leftDepth >= rightDepth) {
            return leftDepth;
        } else {
            return rightDepth;
        }
    }
}
// @lc code=end

File: go/104maximum-depth-of-binary-tree/104.二叉树的最大深度.go
/*
 * @lc app=leetcode.cn id=104 lang=golang
 *
 * [104] 二叉树的最大深度
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxDepth(root *TreeNode) int {
	if root == nil {
		return 0;
	}
	depth := dfs(root, 0)
	return depth
}

func dfs(node *TreeNode, level int) int {
	if node == nil {
		return level
	}
	left := dfs(node.Left, level + 1)
	right := dfs(node.Right, level + 1)
	if left >= right {
		return left
	} else {
		return right
	}
}
// @lc code=end

